[nicholas] iP

nicholastngyk opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://boulderbugle.com/quotes-185mla85))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based

- easy to learn

- ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from here.
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :partying_face:

And it is **FREE**!

Features:

 - [x] Managing tasks
 
 - [ ] Managing deadlines (coming soon)
 
 - [ ] Reminders (coming soon)
 
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
Created hello and goodbye functions

Commit:
Bot will echo user's inputs, and say goodbye when user says bye

Commit:
Created function to add items to a list and print list. Finished level-2.

Commit:
Created a task class to better organise tasks. Additionally, created ability to mark tasks as done and view on list.

Commit:
Moved commands to a separate command class

Commit:
Created a tostring method for tasks

Commit:
Edited text generated by bot

Commit:
Created subclasses for deadlines, events and ToDos with relevant fields added

Commit:
Moved task into a package

Commit:
removed task from original folder

Commit:
Edited previous functions to handle the different types of task.

Commit:
Included short introduction of the bot

Commit:
Removed unnecessary imports

Commit:
Actual seems to equal to expected after Automated Text UI testing

Commit:
Created 2 error handlers, duke exception when the input is empty, and invalid task input when it is not empty, after the task is indicated, it is empty. For example, ( todo )

Commit:
Created delete function

Commit:
Edited text structure

Commit:
Tested with automated text ui testing

Commit:
Created the ability to save tasks in the hard disk when chatbot is closed

Commit:
Edited exceptions to focus primarily on DukeException

Commit:
Accidentally pushed level 7 to main branch. Created level 7 branch and added descriptions to storage class

Commit:
Redefined start and end date of tasks. Changed bot's persona.

Commit:
Updated task classes to take in and print appropriate data and times.

Commit:
Refactored classes into Duke.package

Commit:
Created more OOP by Dividing Classes into Command and task packages.

Commit:
Merge branch 'add-gradle-support'

Commit:
Created test cases for functions and tasks within task list

Commit:
Commented what each functions do

Commit:
Created Filter function for tasklist

Commit:
Edited Java doc

Commit:
Updated list of commands

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
Imported files for checkstyle

Commit:
Edited syntax to follow style guide

Commit:
Merge branch 'master' of https://github.com/nicholastng010601/ip

# Conflicts:
#	README.md

Commit:
Created Main and launcher class to practice GUI.

Commit:
Created UI for chatbot

Commit:
Fixed bug where platform exits without saving tasklist

Commit:
Edited Ui

Originalidk commented on src/main/java/duke/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,56 @@
+        this.ui = ui;
+    }
+
+    /**
+     * Execute the command itself.
+     *
+     * @throws DukeException When there is an error saving.
+     */
```
Should the description for execute be rewritten since the method it overrides has the same purpose? I noticed the same issue in several other places too.

Originalidk commented on src/main/java/duke/utilities/Parser.java:
```diff
@@ -0,0 +1,78 @@
+    /**
+     * An enum to store the various commands for ease of parsing into the respective commands.
+     */
+    enum CommandTypes {
+        BYE,
+        DELETE,
+        LIST,
+        MARK,
+        UNMARK,
+        TODO,
+        DEADLINE,
+        EVENT,
+        HELP,
+        FIND
+    }
```
Good use of enums.

Originalidk commented on src/main/java/duke/utilities/Ui.java:
```diff
@@ -0,0 +1,180 @@
+ */
+public class Ui {
+    protected static Scanner sc;
+    protected String br = "____________________________________________________________\n";
```
Just a nit, but perhaps br could be final instead since you want it to remain constant?
```suggestion
    protected final String br = "____________________________________________________________\n";
```

nicholastngyk replied:
Yes, final would be more appropriate for this instance. Thank you for your comment. Ultimately decided that there was no need for a line break after the ui had been implemented.

Originalidk reviewed (changes requested):
LGTM. Just a few nits to fix.

Commit:
Editted ui

Commit:
Added a title to the chatbot

Commit:
Adhered to coding standards

Commit:
Created a contains function for tasklist

Commit:
Created assertions.

There is currently no way of knowing the current state of the
program until a run time error occurs. All states are assumed.

Assertions was added to keep track of the state of the
program.

Let's add assertions to confirm the state of the program.

Commit:
Merge pull request #1 from nicholastng010601/branch-A-Assertions

Created assertions.

There is currently no way of knowing the current state of the program until a run time error occurs. All states are assumed.

Assertions was added to keep track of the state of the program.

Let's add assertions to confirm the state of the program.

Commit:
Improved code quality

Commit:
Merge pull request #2 from nicholastng010601/branch-A-CodeQuality

Improved code quality

Commit:
Created a function to detect anomalies in events.

The user is able to add events, at times these events might clash.

By creating a detect anomaly function, the chatbot detects possible
event clashes and warns the user that there is a clash, and thus
does not allow the creation of another event over the same time period.

This is done through the creation of an event list class, which stores only
the events, and then compares each event's timing with the new event's
time. Doing so will give the user more control over its events.

Commit:
Deleted false commit. File not in use.

Commit:
Fixed a glitch whereby deleting tasks caused an error.

proto-aiken-13 commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,105 @@
+     * @throws DukeException throws an error when string input is not compatible.
+     */
+    public static String getDescription(String task, String input) throws DukeException {
+        try {
```
Good job on avoiding deep nesting!

proto-aiken-13 commented on src/main/java/duke/utilities/Parser.java:
```diff
@@ -0,0 +1,78 @@
+     * @param ui The ui that prints the messages.
+     * @return the command that was requested by the user.
+     */
+    public static Command parse(String input, TaskList taskList, Storage storage, Ui ui) {
```
Good use of switch cases!

proto-aiken-13 commented on src/main/java/duke/utilities/TaskList.java:
```diff
@@ -0,0 +1,111 @@
+ * A list of tasks the user inputs
+ */
+public class TaskList implements Serializable {
+    protected ArrayList<Task> taskList;
```
This must be a multi-varied name (e.g. name it something like tasks).

proto-aiken-13 reviewed (commented):
Your code is generally very well done and abides to naming conventions of CS2103t quite well, however here are a few issues/highlights from my review of your code.

Commit:
Added more test cases.

Commit:
Merge branch 'branch-B-DetectAnomalies'

# Conflicts:
#	src/main/java/duke/utilities/TaskList.java
#	src/test/java/duke/TaskListTest.java

Commit:
Editted UG

Commit:
Edited readme

Commit:
Followed style guide

Commit:
Fixed a glitch with marking tasks.

Commit:
Modified java doc.

Commit:
Improved user interface of application.

Before, texts were aligned to the right. As a result,
texts printed by the bot were misaligned.

By changing it to the left, it gives users a
better experience by making the text easier
to read.

Commit:
Fixed an error regarding dates of tasks.

Previously, dates that have passed could be set
as deadlines or event timings.

This could both flood the bot and make the inputs
illogical.

Removing this gives the bot the ability to
 prevent illogical dates from being added.

Commit:
Fixed launcher glitch when accessed with windows os

Commit:
no message
